trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/remote/session/RemoteSessionController.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server.remote.session;
+
+import com.google.inject.Inject;
+import io.trino.s3.proxy.server.credentials.Credential;
+import io.trino.s3.proxy.server.remote.RemoteS3Facade;
+import jakarta.annotation.PreDestroy;
+import software.amazon.awssdk.auth.credentials.AwsBasicCredentials;
+import software.amazon.awssdk.auth.credentials.AwsCredentials;
+import software.amazon.awssdk.auth.credentials.AwsSessionCredentials;
+import software.amazon.awssdk.auth.credentials.StaticCredentialsProvider;
+import software.amazon.awssdk.endpoints.Endpoint;
+import software.amazon.awssdk.regions.Region;
+import software.amazon.awssdk.services.sts.StsClient;
+import software.amazon.awssdk.services.sts.auth.StsAssumeRoleCredentialsProvider;
+
+import java.time.Duration;
+import java.util.Map;
+import java.util.Optional;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.Math.toIntExact;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.CompletableFuture.completedFuture;
+
+public class RemoteSessionController
```

Comment:
Should this be an interface? Particularly, if we run multiple instances of the proxy we would need to have some way of sharing sessions amongst them

Reply:
It's a larger question. At some point we'll need to generalize some kind of database/backing store. We have that with `CredentialsController` so maybe this should be an interface too. Any ideas?

Reply:
BTW - one reason this may not be a good candidate for an interface is that `startRemoteSession` and the shutdown hook are common. A session only exists internally for one instance of the proxy.

Reply:
I think we need to formalise the connection between the credentials controller and the session controller. My feeling is that the current design does not have a clearly defined boundary of responsibilities between the `RemoteSessionController` and the `CredentialsController`

The `CredentialsController` needs to get a set of credentials to be used when contacting S3 based on the emulated access credentials, which may be static or STS. These credentials will contain a real credential, which may be STS or static. I think that standardising the contract between the two will make tracking the combination of possibilities clearer?

Reply:
@vagaerg  Makes sense. What is now called `RemoteSessionConfig` should probably come from `CredentialsController` as opposed to being an independent entity. Right?